Jon/feat/req-pay-addr-protocol

Jon-edge opened the pull request:
#### PR Requirements

If you have made **any** visual changes to the GUI. Make sure you have:
- [x] Tested on iOS Tablet
- [x] Tested on small Android
- [ ] n/a

<img src=https://user-images.githubusercontent.com/90650827/165415364-0926fa81-2664-4db5-854f-d9277a2e1e00.png width=300px/>

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201890219170290

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -10,21 +10,147 @@ import { selectWalletForExchange } from '../actions/CryptoExchangeActions.js'
+        if (deepLink.type === 'requestAddress' && deepLink.redir != null) throw new Error(`Invalid 'redir' query in request for payment address`)
+
+        // handle like any other deeplink
+        dispatch(launchDeepLink(parseDeepLink(redir)))
```
This link should bounce out of Edge. launchDeepLink handles links internally in Edge but the most likely use case is to deep link into another app or to a website.

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -10,21 +10,147 @@ import { selectWalletForExchange } from '../actions/CryptoExchangeActions.js'
+        // TODO: Fetch header and proper response error handling, after the POST recipient spec is defined
+        const initOpts = {
+          method: 'POST',
+          headers: { 'Content-Type': 'text/html' },
```
Content-Type should be json

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -10,21 +10,147 @@ import { selectWalletForExchange } from '../actions/CryptoExchangeActions.js'
+    } else {
+      if (redir != null && redir !== '') {
+        // Make sure this isn't some malicious link to cause an infinite redir loop
+        const deepLink = parseDeepLink(redir)
```
`redir` needs to include the parameters that encode the addresses requested by the payer

Jon-edge replied:
In the spec, it shows a BTC address as:
`bitcoin%3A3MdvZezmNMspo5WmLLiBekwkHvKeiMKhgb%3Flabel%3DSatoshi%2520Nakamoto%26message%3DThanks%2520For%2520Freedom`

While an ETH address is:
`ETH_USDC=0x18af1a7713dc3524519a7d1d3eab77828c1202c6`

Is it intentional that only the BTC address is preceeded by `bitcoin:` or should the ethereum address also be preceeded by `ethereum:`?

Jon-edge commented on src/__tests__/DeepLink.test.js:
```diff
@@ -24,6 +24,59 @@ describe('parseDeepLink', function () {
+          { nativeCode: 'ETH', tokenCode: 'UNI' }
+        ],
+        post: undefined,
+        redir: 'testredir.com',
```
In the spec it says for an ETH address the `getaddr` param it's:
`ETH_USDC=0x18af1a7713dc3524519a7d1d3eab77828c1202c6`

But for BTC:
`BTC=bitcoin%3A3MdvZezmNMspo5WmLLiBekwkHvKeiMKhgb%3Flabel%3DSatoshi%2520Nakamoto%26message%3DThanks%2520For%2520Freedom`

Is it intentional for ETH not to have `ethereum:` preceding the address?

paullinator replied:
the prefix is only needed if the address contains params such as a payee and message. This applies across-the-board to Bitcoin and Ethereum.

Jon-edge replied:
This test is just for the parsing, not the client's response after making selections

Jon-edge commented on src/util/CurrencyWalletHelpers.js:
```diff
@@ -30,3 +30,31 @@ export const getAvailableBalance = (wallet: EdgeCurrencyWallet, tokenCode?: stri
+      (tokenCode == null ||
+        (tokenCode != null &&
+          (tokenCode.toUpperCase() === walletNativeCurrencyCode ||
+            enabledTokenCodes.some(enabledTokenCode => tokenCode != null && enabledTokenCode.toUpperCase() === tokenCode.toUpperCase()))))
```
Flow will complain about the last `toUpperCase() === tokenCode.toUpperCase()` statement

paullinator replied:
ok

paullinator commented on src/util/webUtils.js:
```diff
@@ -0,0 +1,44 @@
+  }
+  if (Platform.OS === 'ios') {
+    return SafariView.isAvailable()
+      .then(SafariView.show({ url: uri }))
```
Do not open in SafariView. Just use Linking.openURL

paullinator commented on src/util/DeepLinkParser.js:
```diff
@@ -159,13 +160,14 @@ function parseDownloadLink(url: URL): PromotionLink {
-  const redir = query.redir ?? undefined
+
+  // Ensure redir doesn't end with a '/'
+  const redir = query.redir != null ? enforceLastChar(omitLastChar(query.redir, '/'), '?') : undefined
```
Can't enforce adding a `?` at the end of the redir URI. The redir might already have a param. ie `https://bitwage.com/reqaddr/?requestId=123456`

In that sense, you'd need to add params to the param list and send something like

`https://bitwage.com/reqaddr/?requestId=123456&BTC=1z9823498274987`

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -50,24 +50,22 @@ import { launchDeepLink } from './DeepLinkingActions.js'
-    if (assets.length === 0) throw new Error('No currencies found in request for payment address')
-    if (post == null && redir == null) throw new Error('Post or redir address not found in request for payment address')
-    if (post != null && redir != null) throw new Error('Both post and redir address were specified in request for payment address')
+    if (assets.length === 0) throw new Error(s.strings.bitwage_error_no_currencies_found)
```
Remove reference to `bitwage` this is NOT a bitwage specific feature. It's the Request for Address protocol

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -50,24 +50,22 @@ import { launchDeepLink } from './DeepLinkingActions.js'
-    if (post == null && redir == null) throw new Error('Post or redir address not found in request for payment address')
-    if (post != null && redir != null) throw new Error('Both post and redir address were specified in request for payment address')
+    if (assets.length === 0) throw new Error(s.strings.bitwage_error_no_currencies_found)
+    if ((post == null || post === '') && (redir == null || redir === '')) throw new Error(s.strings.bitwage_error_post_redir)
```
Remove reference to bitwage this is NOT a bitwage specific feature. It's the Request for Address protocol

paullinator commented on src/locales/strings/enUS.json:
```diff
@@ -39,6 +39,14 @@
   "azteco_invalid_code": "Invalid Azteco code.",
   "azteco_service_unavailable": "Error contacting Azteco server. Please try again later.",
   "azteco_btc_only": "Please choose a BTC wallet and scan again.",
+  "bitwage_error_unsupported_chains": "Edge does not support the following chains listed in the request: %s",
```
Remove reference to `bitwage` this is NOT a bitwage specific feature. It's the Request for Address protocol

paullinator commented on src/__tests__/DeepLink.test.js:
```diff
@@ -26,42 +26,79 @@ describe('parseDeepLink', function () {
   describe('requestAddress', () => {
     makeLinkTests({
-      'edge://reqaddr?codes=eth-btc&post=testpost.com': {
+      'edge://reqaddr?codes=eth-btc&post=https://bitwage.com&redir=https://bitwage.com/getaddr&payer=bitwage': {
```
The `post=` and `redir=` params should be uuencoded before being presented as a Request for Address URI. Otherwise they may not parse properly. I realized that I don't have a clear example of this in the docs ☹️

paullinator replied:
The app would then have to uudecode the `post=` and `redir=` params before using them.

paullinator commented on src/__tests__/DeepLink.test.js:
```diff
@@ -26,42 +26,79 @@ describe('parseDeepLink', function () {
+        redir: 'https://bitwage.com/getaddr?',
+        payer: undefined
+      },
+      'edge://reqaddr?codes=ETH_usdc-BTC-BTC_DOGE&redir=https://bitwage.com/getaddr?&payer=bitwage&post=https://bitwage.com': {
```
There are 2 '?' chars in the test URI. This is invalid unless the 2nd one is uuencoded.

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -93,9 +93,10 @@ export const doRequestAddress = async (dispatch: Dispatch, currencyWallets: { [w
     )) {
       const reqNativeCode = asset.nativeCode
       const reqTokenCode = asset.tokenCode
-      const pluginId = getPluginIdFromChainCode(reqNativeCode) ?? '' // Will never by empty string. Already filtered from for loop filter above
```
why was the ?? removed. This can technically still return void

Jon-edge replied:
We already check this with a call to `unsupportedNativeCodes` on line 80. I'll reorganize the logic here

paullinator commented on src/components/modals/TransactionAdvanceDetails.js:
```diff
@@ -40,27 +39,15 @@ export class TransactionAdvanceDetailsComponent extends PureComponent<Props> {
     const { url } = this.props
     if (url == null || url === '') return
-    openBrowserUri(url)
+    openBrowserUri({ uri: url, isSafariView: true })
```
do not use a Safari webview

Jon-edge replied:
This code originally opened SafariView, if available. Was the comment a request to change existing behavior?

[`openBrowserUri`](https://github.com/EdgeApp/edge-react-gui/blob/dacab0259743bb540e4782346f40becb3bf261eb/src/util/WebUtils.js#L14) has a parameter `isSafariView ` to try to use SafariView, only if available.

Jon-edge replied:
This code originally opened SafariView, if available. Was the comment a request to change existing behavior?

[openBrowserUri](https://github.com/EdgeApp/edge-react-gui/blob/dacab0259743bb540e4782346f40becb3bf261eb/src/util/WebUtils.js#L14) has a parameter isSafariView to try to use SafariView, only if available.

paullinator commented on src/util/webUtils.js:
```diff
@@ -2,14 +2,17 @@
-  if (Platform.OS === 'ios') {
+  // Try to open a SafariView, if requested and supported
+  if (isSafariView && Platform.OS === 'ios') {
     return SafariView.isAvailable()
```
You're still using a safariwebview. Just use `openUrl`

paullinator commented on src/util/WebUtils.js:
```diff
@@ -58,3 +57,44 @@ export const stringifyUriAndQuery = (uri: string, query: UriQueryMap): string =>
+ * Parses the query portion of a URL/URI into a UriQueryMap.
+ * Does NOT extract the query from the complete URI!
+ * */
+export const parseQuery = (query?: string): UriQueryMap => {
```
use the `url-parse` library already included in the GUI

Jon-edge replied:
Added TODO

paullinator commented on src/util/WebUtils.js:
```diff
@@ -58,3 +57,44 @@ export const stringifyUriAndQuery = (uri: string, query: UriQueryMap): string =>
+ * -Joins each query with '&'
+ */
+export const stringifyQuery = (query: UriQueryMap): string => {
+  return Object.keys(query)
```
should probably use `url-parse` which does full work of parsing and encoding urls.

Jon-edge replied:
Added TODO

Jon-edge commented:
I rebased the entire feature, fixed it, and re-split/reordered the commits. 

Changes summary:
- Fixed a bug with newly created wallets not getting picked up, not sure if that was pre-existing. Changed doRequestAddress to accept an account so it knows about wallet updates.
- Fixed a bug where it would only detect unsupported native codes. BTC_DOGE should give a warning about being unsupported, but it accepted it. I updated the check to use our new token methods.
- The initial confirmation message used to show the native codes only, making it confusing for displaying tokens. I changed it to show only the token codes. Though the spec only requires a native code with an optional token code, our implementation always populates the token code. ('BTC' translates to 'BTC_BTC')
- Was able to update all but one WebUtil method with url-parse and URLSearchParams.

paullinator commented on src/locales/en_US.js:
```diff
@@ -49,10 +49,22 @@ const strings = {
   alert_dropdown_alert: 'Alert! ',
   alert_dropdown_warning: 'Warning! ',
 
+  and: 'and',
```
This should use string replacement. Also use an easier to search key than `and`

`this_and_that: '%1$s and %2$s'`

Jon-edge replied:
This is used for a string list of potentially 3+ elements, not just two elements. Is this pattern safe translation-wise for 3+ elements?

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -10,21 +10,137 @@ import { selectWalletForExchange } from '../actions/CryptoExchangeActions.js'
+        body: JSON.stringify(jsonPayloadMap)
+      }
+      try {
+        await fetch(post, initOpts)
```
Also need to check for `!response.ok`

```
const response = await fetch(post, initOpts)
if (!response.ok) showError(s.strings.error_contacting_server)
```

paullinator commented on src/actions/ScanActions.js:
```diff
@@ -10,21 +10,137 @@ import { selectWalletForExchange } from '../actions/CryptoExchangeActions.js'
+  }
+}
+
+export const doReturnAddress = async (dispatch: Dispatch, edgeWallet: EdgeCurrencyWallet, link: ReturnAddressLink) => {
```
Can be removed. doReturnAddress is replaced by the V2 spec.

Jon-edge commented on src/util/WebUtils.js:
```diff
@@ -0,0 +1,60 @@
+
+  // Decide whether to use '&' or '?' to add new queries into the uri.
+  const joiningChar = Object.keys(existingQueries).length > 0 ? '&' : '?'
+  return `${omitLastChar(uri, '/')}${joiningChar}${stringifyQuery(query)}`
```
There's a bit of an issue in that the library does not handle query params that are not key=value pairs, hence the special regex replacement in [`stringifyQuery`](https://github.com/EdgeApp/edge-react-gui/blob/6f1d8231c1a758bdb43cf10dd27adbeec8d1758c/src/util/WebUtils.js#L63).

Jon-edge commented on src/util/WebUtils.js:
```diff
@@ -44,20 +42,12 @@ export const openBrowserUri = ({ uri, isSafariView }: OpenBrowserUriParams) => {
  * terminated uri's
  */
-export const stringifyUriAndQuery = (uri: string, query: UriQueryMap): string => {
+export const stringifyUriAndQuery = (uri: string, newQueries: UriQueryMap): string => {
```
There's a bit of an issue in that the library does not handle query params that are not key=value pairs, hence the special regex replacement in [stringifyQuery](https://github.com/EdgeApp/edge-react-gui/blob/6f1d8231c1a758bdb43cf10dd27adbeec8d1758c/src/util/WebUtils.js#L63).

This method leverages that safety mutation in `stringifyQuery`

paullinator commented:
I'm going to approve this but some comments still not addressed such as the SafariWebView issue.
